### ISSUE TITLE
Question about mi_collect

### ISSUE THREAD

**Issue by MasterDuke17** [microsoft/mimalloc#536]
In MoarVM, whenever we do what we call a "full" garbage collection run, we call `malloc_trim()` if it's available. I thought it would make sense to now also call `mi_collect(true)`. However, while testing this out I noticed some odd interaction with heaptrack and our `--full-cleanup` runtime option (which frees everything upon exit, instead of just letting the OS clean up). I use the `--full-cleanup` option when I'm profiling with heaptrack, so it gives an accurate account of leaks (i.e., we try for no leaks with `--full-cleanup). Using a heaptrack that I've added support for mimalloc to, everything looks good normally. However, when I added the `mi_collect(true)` call, now even with `--full-cleanup` heaptrack reports massive leaks. Is there something I should add to my patch to heaptrack to get it to properly handle `mi_collect`?

**Comment by daanx** [comment]
Hmm, that is strange as `mi_collect` only eagerly processes delayed "free"s (from other threads), tries to decommit pages, and reclaims heaps from threads that terminated. Since you hook into `mi_free` / `free`, all of those should already be accounted for.. So really strange that it starts reporting leaks if you add the call ?? 
Do you use multiple threads?  

I cannot explain what is happening; maybe some call is not hooked, but I cannot see how that can be influenced by calling `mi_collect` ??

**Comment by MasterDuke17** [comment]
> Do you use multiple threads?

Yes, there are usually at least two threads running at all times.
